[vLLM][Core]: Add shutdown to vLLM connector

hickeyma opened the pull request:
<!--  Thanks for your contribution to LMCache!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/LMCache/LMCache/blob/dev/CONTRIBUTING.md
2. If this PR closes another issue, add 'Fixes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'Refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

The vLLM connector does not shutdown LMCache gracefully. This because LMCache does not implement the `shutdown` method as defined in the vLLM `KVConnectorBase_V1` interface.  This PR implements the `shutdown` method which calls the cleanup methods for all the objects that are created during startup. 

There are also updates to some of the cleanup methods of the objects themselves to:
- Reduce deadlock and unsafe attribute access
- Add error handling and logging

Changes are as follows:

1. `lmcache/integration/vllm/vllm_v1_adapter.py`
   - Added `shutdown()` method
   - Added timeout protection for all cleanup operations
   - Added missing resource cleanup
   - Added comprehensive logging
   - Added error tracking and reporting

2. `lmcache/v1/cache_engine.py`
   - Enhanced `close()` method with comprehensive logging
   - Enhanced `LMCacheEngineBuilder.destroy()` with detailed logging
   - Added error handling for all operations

3. `lmcache/v1/storage_backend/storage_manager.py`
   - Rewrote `close()` method with timeout protection
   - Added 10-second timeout to thread.join()
   - Added individual backend closure logging
   - Added error handling for event loop stop

4. `lmcache/v1/offload_server/zmq_server.py`
   - Enhanced `process_request()` with exception handling
   - Rewrote `close()` method with timeout
   - Added graceful thread termination
   - Added detailed logging

5. `lmcache/observability.py`
   - Rewrote `LMCacheStatsLogger.shutdown()` method with timeout protection
   - Added smart timeout calculation (log_interval + 2 seconds, minimum 5s)
   - Added comprehensive logging for stats logger shutdown
   - Added error handling for thread join

**Special notes for your reviewers**:

- Need to add/update unit tests

**If applicable**:

- [ ] this PR contains user facing changes - docs added
- [x] this PR contains unit tests

hickeyma committed:
Add shutdown to vLLM connector

The vLLM connector does not shutdown LMCache gracefully. This
commit addsa shutdown method which calls cleanup methods for all
the objects that are created during the startup.

There are also updates to some of the cleanup methods of the objects
themselves to:

- Reduce deadlock and unsafe attribute access
- Add error handling and logging

Changes as follows:

1. `lmcache/integration/vllm/vllm_v1_adapter.py`
   - Added `shutdown()` method
   - Added timeout protection for all cleanup operations
   - Added missing resource cleanup
   - Added comprehensive logging
   - Added error tracking and reporting

2. `lmcache/v1/cache_engine.py`
   - Enhanced `close()` method with comprehensive logging
   - Enhanced `LMCacheEngineBuilder.destroy()` with detailed logging
   - Added error handling for all operations

3. `lmcache/v1/storage_backend/storage_manager.py`
   - Rewrote `close()` method with timeout protection
   - Added 10-second timeout to thread.join()
   - Added individual backend closure logging
   - Added error handling for event loop stop

4. `lmcache/v1/offload_server/zmq_server.py`
   - Enhanced `process_request()` with exception handling
   - Rewrote `close()` method with timeout
   - Added graceful thread termination
   - Added detailed logging

5. `lmcache/observability.py`
   - Rewrote `LMCacheStatsLogger.shutdown()` method with timeout protection
   - Added smart timeout calculation (log_interval + 2 seconds, minimum 5s)
   - Added comprehensive logging for stats logger shutdown
   - Added error handling for thread join

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented on lmcache/integration/vllm/vllm_v1_adapter.py:
```diff
@@ -1126,6 +1126,87 @@ def get_finished(
+    @_lmcache_nvtx_annotate
+    def shutdown(self):
+        # Standard
+        from concurrent.futures import ThreadPoolExecutor, TimeoutError
+        import time
```
Want to use lazy importing instead.

hickeyma commented on lmcache/integration/vllm/vllm_v1_adapter.py:
```diff
@@ -1126,6 +1126,87 @@ def get_finished(
+            _safe_close("lookup_client", self.lookup_client.close, timeout=10.0)
+
+        # Destroy cache engine
+        try:
+            logger.info(f"Destroying LMCache engine: {ENGINE_NAME}")
+            with ThreadPoolExecutor(max_workers=1) as executor:
+                future = executor.submit(LMCacheEngineBuilder.destroy, ENGINE_NAME)
+                try:
+                    future.result(timeout=15.0)
+                    logger.info("LMCache engine destroyed successfully")
+                except TimeoutError:
+                    logger.error(
+                        "Cache engine destroy timed out after 15s. "
+                        "Continuing with shutdown..."
+                    )
+                    errors.append(("cache_engine", "Timeout"))
+        except Exception as e:
+            logger.error(f"Error destroying cache engine: {e}")
+            errors.append(("cache_engine", e))
```
I can pass args and use one safe close function but it would mean that specific tracing and logging would be lost. Hence, going to keep this additional code for better clarity.

hickeyma committed:
Add unit test

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

sammshen reviewed (approved):
LGTM!

hickeyma committed:
Add vllm install to unit tests run on buildkite

vLLM v1 adapter unit test require vllm package.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

ApostaC commented on .buildkite/pipeline.yml:
```diff
@@ -14,6 +14,7 @@ steps:
       uv pip install -r requirements/common.txt
       uv pip install -r requirements/test.txt
+      uv pip install vllm
       uv pip install -e . --no-build-isolation
```
Quick question: why do we need vllm for unit tests? I thought we have vllm integration tests and comprehensive tests to make sure the integration with vllm;
In the meantime, we are also adding unit test at vLLM side to secure the integration

hickeyma replied:
That's a good point. However, the unit tests are testing `LMCacheConnectorV1Impl` in https://github.com/LMCache/LMCache/blob/dev/lmcache/integration/vllm/vllm_v1_adapter.py which imports vLLM packages in https://github.com/LMCache/LMCache/blob/dev/lmcache/integration/vllm/vllm_v1_adapter.py#L8-L39. Is there an alternative to this?

ApostaC replied:
In the long run, we should put those test into vLLM. (Most of vllm_v1_adapter will be put into vllm as well). @sammshen is working on a detailed design proposal for refactoring.

hickeyma replied:
@ApostaC Removed the unit test as I agree its better suited to vLLM level and the test will be introduced in vLLM project shortly.

Also the `LMCacheBuilder` `destroy()` ethos is currently tested in https://github.com/LMCache/LMCache/blob/dev/tests/v1/test_cache_engine.py#L1213-L1407. This makes me think that the issues with shutdown occur when started with vLLM server. Therefore, best tested at vLLM level.

KuntaiDu commented on .buildkite/pipeline.yml:
```diff
@@ -14,6 +14,7 @@ steps:
       uv pip install --upgrade pip setuptools wheel
       uv pip install -r requirements/common.txt
       uv pip install -r requirements/test.txt
+      uv pip install vllm
```
For buildkite unit tests, prefer to not let vLLM be LMCache CI dependency. We are moving integration test to vLLM CI.

hickeyma replied:
Same response as in https://github.com/LMCache/LMCache/pull/1945#discussion_r2534999614

hickeyma replied:
Respondeed in https://github.com/LMCache/LMCache/pull/1945#discussion_r2538998810

hickeyma committed:
Remove unit tests

Removal of unit tests because its a tests at vLLM level and tests are
being introduced into vLLM. Also LMCacheBuilder destroy() method
is tested in unit tests already.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks @ApostaC @KuntaiDu for the reviews and feedback. Updated and ready for review again.

ApostaC reviewed (approved):
LGTM!

hickeyma commented:
@Shaoting-Feng Any ideas why [comprehensive-tests](https://buildkite.com/lmcache/comprehensive-tests/builds/485) is failing? I don't think it is caused by this PR.

ApostaC commented:
@hickeyma Can you update to the latest branch? Shaoting has fixed issues in CI recently.

hickeyma commented:
> Can you update to the latest branch? Shaoting has fixed issues in CI recently.

@ApostaC I did earlier today but I'll try again. Thanks

hickeyma commented:
@KuntaiDu @maobaolong Do you mind taking a look?

hickeyma committed:
Merge branch 'dev' into call-cache-engine-destroy

KuntaiDu reviewed (approved):
LGTM!
